<launch>

<!-- action interface to switch controllers -->
  <node pkg="mechanism_control" type="action_runner" output="screen" />

<!-- Base controller -->
  <param name="base_controller/autostart" type="bool" value="true" />
  <param name="base_controller/odom_publish_rate" value="30.0"/>
  <node pkg="mechanism_control" type="spawner.py" args="$(find pr2_default_controllers)/base_controller.xml" />

<!-- Laser tilt controller -->
  <param name="laser_tilt_controller/autostart" type="bool" value="false" />
  <node pkg="mechanism_control" type="spawner.py" args="--stopped $(find pr2_experimental_controllers)/laser_tilt/laser_tilt_controller.xml" />

<!-- Head controller -->
 <group ns="head_controller" clear_params="true">
    <param name="pan_link_name" type="string" value="head_pan_link" />
    <param name="tilt_link_name" type="string" value="head_tilt_link" />
    <param name="pan_output" type="string" value="head_pan_joint_position_controller" />
    <param name="tilt_output" type="string" value="head_tilt_joint_position_controller" />
 </group>


  <node pkg="mechanism_control" type="spawner.py" args="--stopped $(find pr2_default_controllers)/head_pan_joint_position_controller.xml $(find pr2_default_controllers)/head_tilt_joint_position_controller.xml $(find pr2_default_controllers)/head_position_controller.xml" output="screen" />


<!-- Gripper position controller -->
  <param name="r_gripper_position_controller/autostart" type="bool" value="false" />
  <node pkg="mechanism_control" type="spawner.py" args="--stopped $(find pr2_default_controllers)/r_gripper_position_controller.xml" />

<!-- Gripper effort controller -->
  <node pkg="mechanism_control" type="spawner.py" args="--stopped $(find pr2_default_controllers)/r_gripper_effort_controller.xml" />


<!-- Joint Trajectory controller -->
  <group ns="r_arm_joint_trajectory_controller" clear_params="true">
    <param name="velocity_scaling_factor" value="0.50" type="double" />
    <param name="trajectory_wait_timeout" value="10.0" type="double" />
    
    <param name="r_shoulder_pan_joint/goal_reached_threshold" value="0.25" type="double" />
    <param name="r_shoulder_lift_joint/goal_reached_threshold" value="0.25" type="double" />
    <param name="r_upper_arm_roll_joint/goal_reached_threshold" value="0.25" type="double" />
    <param name="r_elbow_flex_joint/goal_reached_threshold" value="0.25" type="double" />
    <param name="r_forearm_roll_joint/goal_reached_threshold" value="0.25" type="double" />
    <param name="r_wrist_flex_joint/goal_reached_threshold" value="0.25" type="double" />
    <param name="r_wrist_roll_joint/goal_reached_threshold" value="0.25" type="double" /> 
  </group>
  <node pkg="mechanism_control" type="spawner.py" args="--stopped $(find pr2_default_controllers)/r_arm_joint_trajectory_controller.xml" />

<!-- Cartesian wrench controller -->
  <group ns="r_arm_cartesian_wrench_controller" clear_params="true">
    <param name="autostart" type="bool" value="false" />
    <param name="root_name" type="string" value="torso_lift_link" />
    <param name="tip_name" type="string" value="r_gripper_tool_frame" />
  </group>

<!-- Cartesian twist controller -->
  <group ns="r_arm_cartesian_twist_controller" clear_params="true">
    <param name="root_name" type="string" value="torso_lift_link" />
    <param name="tip_name" type="string" value="r_gripper_tool_frame" />
    <param name="output" type="string" value="r_arm_cartesian_wrench_controller" />

    <param name="ff_trans" value="0.0" />
    <param name="ff_rot" value="0.0" />

    <param name="fb_trans/p" value="20.0" />
    <param name="fb_trans/i" value="0.5" />
    <param name="fb_trans/d" value="0.0" />
    <param name="fb_trans/i_clamp" value="1.0" />

    <param name="fb_rot/p" value="0.5" />    <param name="fb_rot/i" value="0.1" />
    <param name="fb_rot/d" value="0.0" />
    <param name="fb_rot/i_clamp" value="0.2" />

    <param name="constraint/active" value="1" />
    <param name="constraint/vel_x" value="0.3" />
  </group>

<!-- Cartesian pose controller -->
  <group ns="r_arm_cartesian_pose_controller" clear_params="true">
    <param name="root_name" type="string" value="torso_lift_link" />
    <param name="tip_name" type="string" value="r_gripper_tool_frame" />
    <param name="output" type="string" value="r_arm_cartesian_twist_controller" />

    <param name="p" value="20.0" />
    <param name="i" value="0.1" />
    <param name="d" value="0.0" />
    <param name="i_clamp" value="0.5" />
  </group>


<!-- Cartesian trajectory controller -->
  <group ns="r_arm_cartesian_trajectory_controller" clear_params="true">
    <param name="root_name" type="string" value="torso_lift_link" />
    <param name="tip_name" type="string" value="r_gripper_tool_frame" />
    <param name="output" type="string" value="r_arm_cartesian_pose_controller" />

    <param name="max_vel_trans" value="0.15" />
    <param name="max_vel_rot" value="0.3" />
    <param name="max_acc_trans" value="0.4" />
    <param name="max_acc_rot" value="0.6" />
  </group>


<node pkg="mechanism_control" type="spawner.py" args="--stopped $(find pr2_default_controllers)/r_arm_cartesian_wrench_controller.xml $(find pr2_default_controllers)/r_arm_cartesian_twist_controller.xml $(find pr2_default_controllers)/r_arm_cartesian_pose_controller.xml $(find pr2_default_controllers)/r_arm_cartesian_trajectory_controller.xml" output="screen" />

</launch>
